DOC-10407 Provide an end to end tutorial on how to troubleshoot lock contention (#18986)
cockroachdb/docs: src/current/v24.3/troubleshoot-lock-contention.md

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-transaction-executions-view.png' | relative_url }}" alt="Transaction Executions view" style="border:1px solid #eee;max-width:100%" />

Depending on when you executed the three transactions, to display the transactions flagged with insights, you may have to select a longer time interval, such as **Past 6 Hours**.

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-time-interval.png' | relative_url }}" alt="Time interval" style="border:1px solid #eee;max-width:100%" />

With an adequate time interval, two [**High Contention**]({% link {{ page.version.version }}/ui-insights-page.md %}#high-contention) insights will be listed for the waiting transactions that experienced contention, **Transaction 2** and **Transaction 3** executed in Example 1.

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-high-contention.png' | relative_url }}" alt="High Contention" style="border:1px solid #eee;max-width:100%" />

### Waiting statement

To identify the exact statement in the transaction that experienced high contention, click the **Latest Transaction Execution ID**

> 452da3e7-e7fa-4801-aa40-c59988d14eb6, the ID of the latest execution with the given [transaction fingerprint]({% link {{ page.version.version }}/ui-transactions-page.md %}).

Reviewer: i think the UI transactions page link should probably go to the subsection at anchor `#transaction-fingerprints-view`?

Scroll to the bottom of the Overview tab to the **Transaction with ID 452da3e7-e7fa-4801-aa40-c59988d14eb6 waited on** section which gives information about the blocking transaction.

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-blocking-transaction.png' | relative_url }}" alt="Blocking transaction" style="border:1px solid #eee;max-width:100%" />

For more information about the blocking transaction, click the **Transaction Fingerprint ID** 08d8b7d97f830463 to open the [**Transaction Details** page]({% link {{ page.version.version }}/ui-transactions-page.md %}#transaction-details-page).

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-blocking-transaction-details.png' | relative_url }}" alt="Blocking transaction details" style="border:1px solid #eee;max-width:100%" />

### Additional practice

For Transaction 3, take steps similar to the above steps for Transaction 2, to identify the waiting statement that experienced high contention and the corresponding blocking transaction.

## Remediate lock contention

### Background context

> Locking conflicts are a natural artifact when business requirements call for concurrent data changes. Realistically, locking conflicts are unavoidable. The locking conflicts, however, are resolved efficiently with regard to the underlying resource utilization. When blocked transactions are waiting on a lock, they are not consuming CPU, disk, or network resources.

Reviewer: I don't really understand what we're trying to say with the latter part of this paragraph

> The locking conflicts, however, are resolved efficiently with regard to the underlying resource utilization. When blocked transactions are waiting on a lock, they are not consuming CPU, disk, or network resources.

this may be technically true but seems irrelevant. the only reason i would be looking at this page is if i am coming here with a problem.  I mean if they're so efficient why do i have to come look at my database's locking docs instead of doing something else?

perhaps there is intent here that I'm missing, but as a reader this seems to say "blocked transactions are efficient, actually" and i'm like "if locking was fine and wasn't being a problem, i wouldn't be looking at these docs"

my suggestion is to drop the last two sentences and just go into the next paragraph, something like

> Locking conflicts are a natural artifact when business requirements call for concurrent data changes. Realistically, locking conflicts are unavoidable.

> Remediation is required when ...

> Remediation is required when locking conflicts are too numerous, resulting in a significant increase in response time and/or decrease in throughput. Remediation of locking conflicts is typically about giving up some functionality in exchange for a reduction in locking contention. [Example 2](#example-2) uses two ways of doing this: [historical queries]({% link {{ page.version.version }}/as-of-system-time.md %}) and a ["fail fast" method]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies). Use these remediations if they fit your application design. Other possible ways to reduce lock conflicts are: [column families]({% link {{ page.version.version }}/column-families.md %}) and [secondary indexes]({% link {{ page.version.version }}/schema-design-indexes.md %}).

Reviewer: > Other possible ways to reduce lock conflicts are: [column families]({% link {{ page.version.version }}/column-families.md %}) and [secondary indexes]({% link {{ page.version.version }}/schema-design-indexes.md %}).

Strongly suggest removing this sentence.  It reads as a checkbox item that is "another way to reduce lock conflicts" but IMO it really needs more explanation if we're going to mention it, it's not at all clear how to go from "i read the column families docs" to "my lock contention is now fixed".  I also don't know offhand what reading through the secondary indexes docs would accomplish re: fixing my lock contention

Finally it's also just confusing to tack it on to this paragraph when this section of the tutorial is not about these things at all

### Historical queries

> In this tutorial's [Example 2](#example-2), Transaction 5 uses a historical query:

Reviewer: forward-referencing a tutorial that the reader hasn't seen yet is pretty confusing 

suggest being less specific here since this is the conceptual overview and adding more overview content.  Tell me why historical queries are a good way to do this, something like

"one way to reduce lock contention is to replace reads with historical reads using AS OF SYSTEM TIME wherever possible. for example, ..."

and then end the section with something like 

"for an example showing this method, see [example 2]() below"

~~~ sql
BEGIN AS OF SYSTEM TIME '-30s';
SELECT * FROM t2 WHERE k=4; -- historical read
COMMIT;
~~~

Consider the following when using [historical queries]({% link {{ page.version.version }}/as-of-system-time.md %}):

> - Use historical queries only if the application can use data that is 5 seconds old or older.

Reviewer: Strongly recommend not hardcoding this "5 seconds" and instead linking to the definition of `follower_read_timestamp` in https://www.cockroachlabs.com/docs/stable/as-of-system-time#parameters which already has the timing information (it will probably never get updated here when the follower read lag gets updated because this usage will not show up in a search for "follower_read_timestamp" or "follower read")

- Historical queries primarily benefit read-only transactions.
- Historical queries operate below [closed timestamps]({% link {{ page.version.version }}/architecture/transaction-layer.md %}#closed-timestamps) and therefore have perfect concurrency characteristics - they never wait on anything and never block anything.

> - Historical queries have the best possible performance, since they are served by the nearest replica.

Reviewer: suggest linking 'replica' to https://www.cockroachlabs.com/docs/v24.3/architecture/glossary.html#replica

> ### &quot;fail fast&quot; method

Reviewer: section title should be capitalized "Fail fast"

> In this tutorial's [Example 2](#example-2), Transaction 6 uses a &quot;fail fast&quot; method:

Reviewer: same comment here as above re: write more overview about _why_ this is a good method, and then later in the section link to [Example 2]()

Author: added more overview

~~~ sql
BEGIN;
SELECT * FROM t2 WHERE k=4 FOR UPDATE NOWAIT; -- fail fast write
UPDATE t2 SET v=4034 WHERE k=4;
COMMIT;
~~~

> Consider the following when using a ["fail fast" method]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies):

Reviewer: this is confusingly worded, since i thought this was the 'fail fast' section but this is linking to the SELECT FOR UPDATE doc which doesn't say anything about "fail fast"

suggest revising this to make clear it's an external link to the statement docs, something like

> Consider the following when using  a "fail fast" method such as [`SELECT FOR UPDATE ... NOWAIT`]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies):

Author: removed this sentence and updated the section

> - &quot;Fail fast&quot; could be a reasonable protective measure in the application to handle "hot update key" situations, for example, when an application needs to be able to handle an arbitrary large surge of updates on the same key.

Reviewer: >  arbitrary large surge of updates

nit: should be "arbitrarily large"

Author: corrected

- The most direct method of "failing fast" is using pessimistic locking with [SELECT FOR UPDATE … NOWAIT]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies). It can reduce or prevent failures late in a transaction's life (e.g. at the `COMMIT` time), by returning an error early in a contention situation if a row cannot be locked immediately.

> - A more &quot;buffered fail fast&quot; approach would be to control the maximum length of a lock wait-queue that requests are willing to enter and wait in, with the cluster setting  [`kv.lock_table.maximum_lock_wait_queue_length`](https://github.com/cockroachdb/cockroach/pull/66146) (default: `0`). It can provide some level of quality-of-service with a response time predictability in a severe per-key contention. If set to a non-zero value and an existing lock wait-queue is already equal to or exceeding this length, requests will be rejected eagerly instead of entering the queue and waiting.

Reviewer: this is a non-public cluster setting that is undocumented

if we're not going to the trouble of flipping that cluster setting to public (a 1-line code change) we should definitely not be adding it to the docs

we should also DEFINITELY not be linking to a github PR as the reference docs for something that has pretty deep consequences for cluster operation

Author: removed

Author: removed

### Initial Data for Example 2

In any of the SQL shells, create a second table and insert some data:

{% include_cached copy-clipboard.html %}
~~~ sql
DROP TABLE IF EXISTS t2;
CREATE TABLE t2 (k INT PRIMARY KEY, v INT);
INSERT INTO t2 VALUES (4,4), (5,5), (6,6);
~~~

### Example 2

> In this example, Transaction 4 is a write that does not block either Transaction 5, a read, or Transaction 6, a write. Transaction 4 locks key `k=4`. When Transaction 5 tries to read key `k=4`, it does not experience lock contention because it does not have to wait for the lock on the key to be released. Transaction 5 uses [`AS OF SYSTEM TIME`]({% link {{ page.version.version }}/as-of-system-time.md %}) to do a historical read. When Transaction 6 executes the [`SELECT ... FOR UPDATE NOWAIT`]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies) on key `k=4`, an error is returned since the key `k=4` cannot be locked immediately. In other words, Transaction 6 "fails fast". It does not even attempt to do an `UPDATE` write to key `k=4`, so it does not experience lock contention.

Reviewer: suggest adding more overview here before jumping right into the gory details, something like "in this example, we will show how to do ${HIGH LEVEL DESCRIPTION OF THE EXAMPLE}.  Transaction 4 is a write that does not block ..."

Author: added more overview

Transaction 4 (blocking write)   | Transaction 5 (historical read) | Transaction 6 (fail fast write)
---------------------------------|------------------------------|--------------------------------
`BEGIN;`                         |                              |
`UPDATE t2 SET v=4014 WHERE k=4;`| `BEGIN AS OF SYSTEM TIME '-30s';`|
lock k=4                         | `SELECT * FROM t2 WHERE k=4;`| `BEGIN;`
                                 | not waiting                  | `SELECT * FROM t2 WHERE k=4 FOR UPDATE NOWAIT;`
                                 | k=4,v=4                      | error: could not obtain lock
`COMMIT;`                        |                              | `UPDATE t2 SET v=4034 WHERE k=4;`
success, k=4,v=4014              |                              | not waiting
                                 | `COMMIT;`                    | `COMMIT;`
                                 | success                      | failure                                 
                                 |                              | `SELECT * FROM t2 WHERE k=4;`
                                 |                              | k=4, v=4014

### SQL statements for Example 2

> To reproduce Example 2 in CockroachDB to show how to remediate a waiting read and write experiencing lock contention, execute the following SQL statements in the given order in the specified terminal.

Reviewer: suggest shortening this a bit to something like

> To reproduce [Example 2](LINK), execute the following SQL statements in the given order in the specified terminal.

Author: shortened sentence

~~~

Transaction 5 does a historical read and should output the following:

~~~
  k | v
----+----
  4 | 4
(1 row)
~~~

**Terminal 3**

{% include_cached copy-clipboard.html %}
~~~ sql
SET application_name = 'Transaction 6';  -- to distinguish between transactions
BEGIN;
SELECT * FROM t2 WHERE k=4 FOR UPDATE NOWAIT; -- fail fast write
UPDATE t2 SET v=4034 WHERE k=4;
~~~

The `SELECT ... FOR UPDATE NOWAIT` returns:

~~~
ERROR: could not obtain lock on row (k)=(4) in t2@t2_pkey
SQLSTATE: 55P03
ERROR: current transaction is aborted, commands ignored until end of transaction block
SQLSTATE: 25P02
~~~

**Terminal 1**

{% include_cached copy-clipboard.html %}
~~~ sql
COMMIT;
~~~

**Terminal 2**

{% include_cached copy-clipboard.html %}
~~~ sql
COMMIT;
~~~

**Terminal 3**

`COMMIT` Transaction 6. Since the `SELECT` statement in Transaction 6 generated an error, `COMMIT` is equivalent to `ROLLBACK`, which aborts the transaction and discards the `UPDATE`. Afterward, verify that the `UPDATE` was discarded.

{% include_cached copy-clipboard.html %}
~~~ sql
COMMIT;
SELECT * FROM t2 WHERE k=4;
~~~

The `SELECT` statement should output the following:

~~~
  k |  v
----+-------
  4 | 4014
(1 row)

> ~~~

Reviewer: This page [needs a 'See also' section](https://github.com/cockroachdb/docs/blob/main/StyleGuide.md#see-also), easiest way to populate is with the links you were using in the page above

Author: added

summary: This tutorial presents how to understand lock contention, how to identify waiting and blocking transactions, and possible ways to remediate lock contention.
toc: true
---

Lock contention is a type of [transaction contention]({% link {{ page.version.version }}/performance-best-practices-overview.md %}#transaction-contention) that occurs when a [transaction]({% link {{ page.version.version }}/transactions.md %}) is unable to complete due to another concurrent or recent transaction attempting to write to the same data. Lock contention may be the cause of [slow SQL query performance]({% link {{ page.version.version }}/eventlog.md %}#sql-slow-query-log).

This tutorial presents:

- How to [understand lock contention](#understand-lock-contention) by reproducing a basic example.
- How to [identify waiting and blocking transactions](#identify-waiting-and-blocking-transactions) by using the Insights page of the DB Console.
- Possible ways to [remediate lock contention](#remediate-lock-contention).

## Before you begin

[Download]({% link releases/index.md %}) and [Install CockroachDB]({% link {{ page.version.version }}/install-cockroachdb-mac.md %}).

### Terminal 1

> The examples in this tutorial will use three terminals, one for each transaction.

Reviewer 2: It seems like this line "The examples in this tutorial..." could precede the "Terminal 1" heading. Since it is really about all of the following sections.

In the first terminal, use the [`cockroach demo`]({% link {{ page.version.version }}/cockroach-demo.md %}) command to start a temporary, in-memory CockroachDB cluster of one node.

{% include_cached copy-clipboard.html %}
~~~ shell
cockroach demo --no-example-database
~~~

It will open an interactive SQL shell to the cluster set to an empty database called `defaultdb`. This database is used for testing and some internal databases.

> To distinguish each terminal set the `application_name`. In the SQL shell in the first terminal, execute:

Reviewer 2: I wonder if the precise directions that we're in the SQL shell and the first terminal are needed here when the preceding line and section establishes this.
```suggestion
To distinguish each terminal that you'll use in this tutorial, set the `application_name`:
```

{% include_cached copy-clipboard.html %}
~~~ sql

> SET application_name = 'Transaction 1';  -- to distinguish between transactions

Reviewer 2: I'm not sure that the comment in the code is also necessary?

~~~

> To [connect additional SQL clients to the demo cluster]({% link {{ page.version.version }}/cockroach-demo.md %}#connect-an-additional-sql-client-to-the-demo-cluster), in the SQL shell use `\demo ls` to list the connection parameters in the demo cluster:

Reviewer 2: ```suggestion
To [connect additional SQL clients to the demo cluster]({% link {{ page.version.version }}/cockroach-demo.md %}#connect-an-additional-sql-client-to-the-demo-cluster), run:
```

Author: updated

{% include_cached copy-clipboard.html %}
~~~ sql
\demo ls
~~~

> The output will be similar to:

Reviewer 2: ```suggestion
The output will list the connection parameters in the demo cluster:
```

Author: updated

   (cli)      cockroach sql --certs-dir=/Users/myuser/.cockroach-demo -u demo -d defaultdb
~~~

### Terminal 2

In a second terminal, open another SQL shell to the demo cluster using the `cli` command from the `\demo ls` output:

{% include_cached copy-clipboard.html %}
~~~ shell
cockroach sql --certs-dir=/Users/myuser/.cockroach-demo -u demo -d defaultdb
~~~

In the second SQL shell, set `application_name`:

{% include_cached copy-clipboard.html %}
~~~ sql
SET application_name = 'Transaction 2';  -- to distinguish between transactions
~~~

### Terminal 3

In a third terminal, open another SQL shell to the demo cluster using the `cli` command from the `\demo ls` output:

{% include_cached copy-clipboard.html %}
~~~ shell
cockroach sql --certs-dir=/Users/myuser/.cockroach-demo -u demo -d defaultdb
~~~

In the third SQL shell, set `application_name`:

{% include_cached copy-clipboard.html %}
~~~ sql
SET application_name = 'Transaction 3';  -- to distinguish between transactions
~~~

### DB Console

In a web browser, open the DB Console to the demo cluster using the `webui` address from the `\demo ls` output:

~~~
http://127.0.0.1:8080/demologin?password=demo28806&username=demo
~~~

> ## Understand lock contention

Reviewer 2: It seems like each H2 section is a step that should be run sequentially as a procedural (The next step is dependent on having completed the last). You could consider Having ## Step 1. Understand lock contention, ## Step 2. Identify waiting and blocking transactions... and so on

Author: Added "Step N" to H2 sections.

> 

Author: Added the overarching sentence.

~~~
  k |  v
----+-------
  2 | 2012
(1 row)
~~~

`COMMIT` Transaction 2:

{% include_cached copy-clipboard.html %}
~~~ sql
COMMIT;
~~~

**Terminal 3**

`COMMIT` Transaction 3 and verify that the `UPDATE` has succeeded:

{% include_cached copy-clipboard.html %}
~~~ sql
COMMIT;
SELECT * FROM t where k=2;
~~~

The `SELECT` statement should output the following:

~~~
  k |  v
----+-------
  2 | 2032
(1 row)
~~~

## Identify waiting and blocking transactions

This section of the tutorial uses the [**Insights** page]({% link {{ page.version.version }}/ui-insights-page.md %}#transaction-executions-view) of the DB Console to identify waiting and blocked transactions in the demo cluster. With a CockroachDB {{ site.data.products.cloud }} cluster, the {{ site.data.products.cloud }} Console has a similar [**Insights** page]({% link cockroachcloud/insights-page.md %}#transaction-executions-view). You can also use the [`crdb_internal`]({% link {{ page.version.version }}/performance-recipes.md %}#identify-transactions-and-objects-that-experienced-lock-contention) system catalog to view tables and indexes that experienced contention.

> This section of the tutorial assumes you have already run the SQL statements from [Example 1](#example-1). When troubleshooting lock contention in your own workload, you can adapt the following steps using the DB Console or the {{ site.data.products.cloud }} Console.

Reviewer 2: Really small nit, it would be great to rephrase the start of one of these paragraphs to avoid the double "This section..." but not a huge deal.

Author: updated

On the **Transaction Execution** page, navigate to the **Statement Executions** tab. In the list of statement executions, in the **Insights** column for `SELECT * FROM t where k = _`, there should be the **High Contention** insight. In [Example 1](#example-1), *Transaction 2* had one statement (other than `SHOW database`). In a transaction with multiple statements, use this page to pinpoint the exact statement that experienced high contention.

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-waiting-statement.png' | relative_url }}" alt="Waiting statement" style="border:1px solid #eee;max-width:100%" />

### Blocking transaction

To identify the transaction that blocked **Transaction 2** and caused it to experience high contention, navigate back to the **Overview** tab.

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-overview-tab.png' | relative_url }}" alt="Overview tab" style="border:1px solid #eee;max-width:100%" />

> Scroll to the bottom of the Overview tab to the **Transaction with ID ... waited on** section which gives information about the blocking transaction.

Author: updated

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-blocking-transaction.png' | relative_url }}" alt="Blocking transaction" style="border:1px solid #eee;max-width:100%" />

For more information about the blocking transaction, click the **Transaction Fingerprint ID** to open the [**Transaction Details** page]({% link {{ page.version.version }}/ui-transactions-page.md %}#transaction-details-page).

<img src="{{ 'images/v24.3/troubleshoot-lock-contention-blocking-transaction-details.png' | relative_url }}" alt="Blocking transaction details" style="border:1px solid #eee;max-width:100%" />

### Additional practice

> For [*Transaction 3*](#example-1), take steps similar to the above steps for *Transaction 2*, to identify the waiting statement that experienced high contention and the corresponding blocking transaction.

Author: updated

## Remediate lock contention

### Background context

Locking conflicts are a natural artifact when business requirements call for concurrent data changes. Realistically, locking conflicts are unavoidable.

> Remediation is required when locking conflicts are too numerous, resulting in a significant increase in response time and/or decrease in throughput. Remediation of locking conflicts is typically about giving up some functionality in exchange for a reduction in locking contention. [Example 2](#example-2) uses two ways of doing this: [historical queries]({% link {{ page.version.version }}/as-of-system-time.md %}) and a ["fail fast" method]({% link {{ page.version.version }}/select-for-update.md %}#wait-policies). Use these remediations if they fit your application design.

Author: updated
